Documentation/proposals/202310-automated-sharding.md

```diff
@@ -0,0 +1,164 @@
+
+* Eventual cardinality explosions have a big blast radius if a single Prometheus instance is responsible for scraping the majority of monitored applications.
+* Recovering from a crash takes several minutes due to the WAL replay.
+
+Meanwhile, another strategy would be to use Horizontal Pod Autoscalers (HPAs) with Prometheus statefulsets by increasing the number of replicas. However, arguably it makes things more complicated:
+
+* Replicas with the same scrape configuration will not shard the scraping load into different instances. Instead, we get several Prometheus instances generating duplicated metrics.
+
+# Goals
+
+* Enable scaling of the Prometheus shards up and down via Horizontal Pod Autoscaler objects.
+
+# Non-Goals
+
+* Implement the autoscaling logic into the Prometheus operator.
+* Automatic balancing of targets to ensure homogeneous load across all Prometheus pods.
+
+# Audience
+
+* Users looking for a shard autoscaling mechanism with a good cost-reliability ratio.
+
+# How
+
+Today, there are a few strategies to measure load on top of Prometheus instances.
+
+1. Perhaps you can use [Kubernetes Metrics API](https://kubernetes.io/docs/tasks/debug/debug-cluster/resource-metrics-pipeline/#metrics-api) to measure load in terms of CPU and/or Memory usage.
+
+2. Another strategy might be to measure the amount of samples ingested per second
+
+- Prometheus Agents: `rate(prometheus_agent_samples_appended_total[5m])`
+- Prometheus Servers: `rate(prometheus_tsdb_head_samples_appended_total[5m])`
+
+3. Or even measuring samples queried per second for Prometheus instances under heavy query load.
+
+- Prometheus Servers: `rate(prometheus_engine_query_samples_total[5m])`
+
+Modern Custom Horizontal Pod Autoscalers (HPAs), such as Keda are already capable of [scaling CRDs](https://keda.sh/docs/2.11/concepts/scaling-deployments/#scaling-of-custom-resources) based on different parameters, such as [CPU, Memory](https://keda.sh/docs/2.11/scalers/cpu/) or even [custom Prometheus metrics](https://keda.sh/docs/2.11/scalers/prometheus/).
+
+## Scale Status subresource
+
+When working with CRDs, common HPAs (e.g. Keda) depends on a Status subresource called [Scale](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#scale-subresource). Unfortunately, Prometheus-Operator resources still don't implement this field. There was an attempt to implement this in https://github.com/prometheus-operator/prometheus-operator/pull/4735 but was reverted because the PR was scaling replicas instead of shards. Although scaling replicas can help with high availability, it is expensive and hard to manage since it duplicates scrapes while not reducing the load on top of Prometheus because all replicas still scrape the same targets. Sharding serves as a better autoscaling strategy since the load is spread into all instances instead of getting duplicated.
```

Comment:
I would omit the history.
```suggestion
When working with CRDs, common HPAs (e.g. Keda) depends on a subresource called [Scale](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#scale-subresource). This proposal suggests to implement the scale subresource for the Prometheus and PrometheusAgent CRDs. Instead of working on the "replicas" count, it will operate on the "shards" count because the purpose of scaling up (resp. down) is to distribute the same number of targets across more (resp. less) Prometheus instances.
```